Fix the certificate rotation threshold and add jitter.

jcbsmpsn opened the pull request:
Adjusts the certificate rotation threshold to be fixed, with some jitter to
spread out the load on the Certificate Signing Request API. The rotation
threshold is fixed at 20% now, meaning when 20% of the certificate's total
duration is remaining, the certificate manager will attempt to rotate, with
jitter +/-10%. For certificates of duration 1 month that means they will
rotate after 24 days, +/- 3 days.

On a 6000 node cluster, assuming all nodes added at nearly the same time, this
should result in 6000 nodes rotating spread over 6 days (total range of the
jitter), or ~42 nodes / hour requesting new certificates.

jcbsmpsn committed:
Fix the certificate rotation threshold and add jitter.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42498)
<!-- Reviewable:end -->

jcbsmpsn commented:
@dchen1107 @wojtek-t Addressing your concerns from [this PR](https://github.com/kubernetes/kubernetes/pull/41912)
about the scaling impacts of certificate rotation in the kubelet.

This will spread out the activity as described in the PR description.

mikedanese commented on pkg/kubelet/certificate/certificate_manager.go:
```diff
@@ -160,9 +155,14 @@ func (m *manager) shouldRotate() bool {
+	// cluster creation time), they won't all try to rotate certificates at the
+	// same time for the rest of the life of the cluster.
+	jitter := (rand.Float64() * 20) - 10
+	return remaining < 0 || (remaining/total*100) < (shouldRotatePercent+jitter)
```
can you just use https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Jitter

jcbsmpsn replied:
Done.

mikedanese commented on pkg/kubelet/certificate/certificate_manager.go:
```diff
@@ -160,9 +154,19 @@ func (m *manager) shouldRotate() bool {
+	// For the parameters of the Jitter function:
+	//    [total*0.7, total+total*0.7*0.28] ~ [total*0.7, total*0.9]
+	jitteryDuration := wait.Jitter(time.Duration(totalDuration*0.7), 0.28)
+	rotationThreshold := m.cert.Leaf.NotBefore.Add(jitteryDuration)
```
I think the distribution will be a lot closer to 70% if you repeatedly redraw the rotationThreshold. Should you be saving it? How often is shouldRotate called?

jcbsmpsn replied:
About once an hour. Rather than saving the threshold I intend to modify the loop in a follow on PR so that it doesn't check periodically but instead just pauses for the correct duration.

wojtek-t replied:
I'm fine with both options (saving the result or sleeping for that long), but I agree with Mike that we shouldn't leave it like that.

BTW - I have a question: how often we will be usually revoking those certificates? In other word, assuming that we have X nodes, how many per second/minute we will be restarting?

jcbsmpsn replied:
We won't be revoking certificates, just handling certificate expiration. That maybe the scenario you were actually considering but I want to clarify because some people are looking for actual certificate revocation.

The duration of a certificate is currently 5 years I think, and is configurable with a cluster setting. When this is released, we'll probably want to shorten that, maybe every few months. We have no guidance on that yet.

We have a plan to not restart at all. It depends on a new feature in Golang 1.8, [GetClientCertificate](https://golang.org/doc/go1.8#crypto_tls). GetClientCertificate was [broken](https://github.com/golang/go/issues/19264) in the 1.8 release and will be fixed in Golang 1.8.1.

ericchiang replied:
The API server doesn't have code to revoke certificates today. There's been some provisional work to introduce this (https://github.com/kubernetes/kubernetes/pull/33519), but it's still not clear how we'd build in a remote CRL check.

> The duration of a certificate is currently 5 years I think, and is configurable with a cluster setting. When this is released, we'll probably want to shorten that, maybe every few months. We have no guidance on that yet.

Is this in the kubelet code or the certificate signer controller? We should definitely make it configurable.

jcbsmpsn commented:
@k8s-bot gce etcd3 e2e test this
@k8s-bot bazel test this

jcbsmpsn commented:
@dchen1107 @wojtek-t Could you have a look at this?

wojtek-t commented:
> @dchen1107 @wojtek-t Could you have a look at this?

Sorry was on Kubecon for last couple of days. Will try to get to this later today or on Monday.

wojtek-t commented on pkg/kubelet/certificate/certificate_manager.go:
```diff
@@ -160,9 +148,19 @@ func (m *manager) shouldRotate() bool {
+	//
+	// For the parameters of the Jitter function:
+	//    [total*0.7, total+total*0.7*0.28] ~ [total*0.7, total*0.9]
+	jitteryDuration := wait.Jitter(time.Duration(totalDuration*0.7), 0.28)
```
I'm lost in this.
If I understand correctly, what you would like to achieve is that:
- certificate created at t1, expires at t2 (dt := t2 - t1)
- you would like it to be refreshed not earlier than t1 + 0.7dt, and not later than t1 + 0.9dt
Is that right?

If so, this should be:
```
jitteryDuration := wait.Jitter(time.Duration(totalDuration*0.2), 1.0) + time.Duration(totalDuration*0.7)
```

jcbsmpsn replied:
Yes, that's what I want. I think yours is better than what I came up with. Although, if I understand the [documentation for Jitter](https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Jitter) correctly, it should be something like:

```go
jitteryDuration := wait.Jitter(time.Duration(totalDuration*0.2), 1.0) + time.Duration(totalDuration*0.5)
```

as `wait.Jitter(time.Duration(totalDuration*0.2), 1.0)` will produce a value between 0.2 of the original value and 0.4 of the original value.

wojtek-t commented on pkg/kubelet/certificate/certificate_manager.go:
```diff
@@ -160,9 +148,16 @@ func (m *manager) shouldRotate() bool {
+	// jitter, if a number of nodes are added to a cluster at approximately the
+	// same time (such as cluster creation time), they won't all try to rotate
+	// certificates at the same time for the rest of the life of the cluster.
+	jitteryDuration := wait.Jitter(time.Duration(totalDuration*0.2), 1.0) + time.Duration(totalDuration*0.5)
```
Shoudn't it be: s/0.5/0.7/ ?
I think that now it is producing together a number in 0.5-0.7 (random from [0..0.2] + 0.5).

jcbsmpsn replied:
I think 0.5 is correct. My reasoning is, [Jitter documentation](https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Jitter):

    Jitter returns a time.Duration between duration and duration + maxFactor * duration.

So it returns a value between the first parameter (0.2) and the first parameter + (first parameter * second parameter). So, the result of Jitter should be in the range of [0.2, 0.2 + (0.2 * 1)] = [0.2, 0.4].

wojtek-t replied:
aah ok. You're right this is correct. Though in my opinion it is misleading for the user. Personally I would change it to:
```
wait.Jitter(totalDuration, 0.2) + time.Duration(totalDuration*0.7)
```
It's shorter and I think easier to understand.

jcbsmpsn replied:
That would result in jitter in the range of [1, 1 + (1 * 0.2)] = [1, 1.2], so the jitter alone would be beyond the expiration of the certificate.

wojtek-t replied:
Yeah sure. I guess i wanted to write:
```
wait.Jitter(totalDuration, 0.2) - time.Duration(totalDuration*0.3)
```
then it will be within [1, 1.2] - 0.3 = [0.7, 0.9]

jcbsmpsn commented:
@k8s-bot non-cri e2e test this

jcbsmpsn commented:
/assign @vishh

mikedanese commented:
/lgtm

vishh commented:
/approve

jcbsmpsn commented:
/label release-note-none

jcbsmpsn commented:
/release-note-none
